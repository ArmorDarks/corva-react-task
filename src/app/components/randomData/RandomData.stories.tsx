import React from 'react'

import RandomData from './RandomData'

export default {
  component: RandomData,
  title: 'RandomData'
}

const seriesData = [{
  value: -59.168341304150964,
  timestamp: 1570474691564
}, {
  value: -81.39908194859946,
  timestamp: 1570474694403
}, {
  value: 90.20003681572823,
  timestamp: 1570474697507
}, {
  value: -36.39152397661461,
  timestamp: 1570474700740
}, {
  value: -87.92854780324105,
  timestamp: 1570474707417
}, {
  value: -78.59754943742128,
  timestamp: 1570474711327
}, {
  value: -54.398113089246976,
  timestamp: 1570474712735
}, {
  value: 20.69819072092662,
  timestamp: 1570474718902
}, {
  value: -37.775509709049885,
  timestamp: 1570474721594
}] as const

export const initial = () =>
  <RandomData
    randomData={seriesData}
    updateThreshold={(() => {}) as any}
  />

export const withIsConnected = () =>
  <RandomData
    isConnected
    randomData={seriesData}
    updateThreshold={(() => {}) as any}
  />

export const withThreshold = () =>
  <RandomData
    isConnected
    randomData={seriesData}
    threshold={15}
    updateThreshold={(() => {}) as any}
  />
