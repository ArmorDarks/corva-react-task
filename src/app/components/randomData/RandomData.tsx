/* eslint no-unused-vars: "off" */
/* eslint @typescript-eslint/no-unused-vars: "error" */

import React from 'react'

import { BarChart, SplineChart } from '..'
import { RandomData as RandomDataContract } from '../../../contracts/RandomData'
import { Threshold, updateThreshold } from '../../store/randomData'

const CHART_MAX_SERIES = 10

export interface Props {
  isConnected?: boolean,
  randomData: readonly RandomDataContract[],
  threshold?: Threshold,
  updateThreshold: typeof updateThreshold,
}

/**
 * Display Random Data charts and some additional controls
 */
const RandomData: React.FC<Props> = ({ isConnected, randomData, threshold, updateThreshold }) => {
  return (
    <div className='h-margin-top+'>

      <div className='Wrapper h-flex h-flex--wrap h-flex-x--between h-flex-y--center h-margin-ends+'>

        <div className='h-flex h-flex--wrap h-flex-y--center'>
          <h1 className='h-inline h-margin-ends'>Random Data Stats</h1>

          <div className='h-margin-left h-margin-ends'>
            {isConnected ? 'online' : 'offline'}
          </div>

        </div>

        <div className='h-margin-ends'>
          <label className='h-text- h-margin-right-' htmlFor='threshold'>Threshold</label>
          <input
            type='number'
            id='threshold'
            name='threshold'
            onChange={(e) => updateThreshold(e.target.value === '' ? null : +e.target.value)}
            value={threshold === null ? '' : threshold}
          />
        </div>

      </div>

      <div className='o-grid'>

        <div className='o-grid__iterm h-1/2'>
          <BarChart
            seriesName='Random data occurrences'
            seriesData={randomData.map(({ value }) => value)}
          />
        </div>

        <div className='o-grid__iterm h-1/2'>
          <SplineChart
            seriesName='Random data'
            seriesData={randomData.map((data) => [data.timestamp, data.value] as const)}
            threshold={threshold === null ? undefined : threshold}
            limit={CHART_MAX_SERIES}
          />
        </div>

      </div>

    </div>
  )
}

export default RandomData
