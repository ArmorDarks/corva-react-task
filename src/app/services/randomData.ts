/* eslint no-unused-vars: "off" */
/* eslint @typescript-eslint/no-unused-vars: "error" */

import io from 'socket.io-client'
import { RANDOM_DATA_CHANNEL, RandomData } from '../../contracts/RandomData'

type OnResponse = (data: RandomData) => any

/**
 * Factory for RandomData socket connector. Mostly need for the ability to inject mock client in tests.
 * @param client Instance of the 'socket.io-client'
 * @param url Url of the server
 * @param port Port of the server
 */
export const SubscribeOnRandomData = (
  client: typeof io,
  url: string,
  port: string
) => (onResponse: OnResponse) =>
  client(`${url}:${port}`)
    .on(RANDOM_DATA_CHANNEL, onResponse)

/**
 * Connect to RandomData provider via sockets and receive messages in callback
 * @param onResponse Callback to be invoked when new data received
*/
export default (onResponse: OnResponse) => {
  if (!process.env.URL) {
    throw new Error('[services/randomData] sockets server `URL` environment variable is not specified')
  }

  if (!process.env.PORT) {
    throw new Error('[services/randomData] sockets server `PORT` environment variable is not specified')
  }

  SubscribeOnRandomData(io, process.env.URL, process.env.PORT)(onResponse)
}
