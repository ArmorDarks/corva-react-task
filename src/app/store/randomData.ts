/* eslint no-unused-vars: "off" */
/* eslint @typescript-eslint/no-unused-vars: "error" */

import { Dispatch, Reducer } from 'redux'
import { RandomData, RandomDataValue } from '../../contracts/RandomData'
import { assertNever } from '../utils/assertNever'

export const INITIAL_THRESHOLD: RandomDataValue = 15

// ====================================
// State
// ====================================

export interface State {
  readonly data?: readonly RandomData[]
  readonly threshold: RandomDataValue
}

export const initState: State = {
  data: [],
  threshold: INITIAL_THRESHOLD
}

// ====================================
// Action types
// ====================================

export const UPDATE_DATA = 'app/data/UPDATE_DATA'

export interface UpdateData {
  readonly type: typeof UPDATE_DATA;
  readonly payload: RandomData
}

export const UPDATE_THRESHOLD = 'app/data/UPDATE_THRESHOLD'

export interface UpdateThreshold {
  readonly type: typeof UPDATE_THRESHOLD;
  readonly payload: RandomDataValue
}

export type Action =
  UpdateData |
  UpdateThreshold

// ====================================
// Action creators
// ====================================

export const createUpdateData = (payload: RandomData): UpdateData => {
  return {
    type: UPDATE_DATA,
    payload
  }
}

export const createUpdateThreshold = (payload: RandomDataValue): UpdateThreshold => {
  return {
    type: UPDATE_THRESHOLD,
    payload
  }
}

// ====================================
// Actions
// ====================================

export const updateData = () => (dispatch: Dispatch<Action>): Promise<UpdateData> => {
  return Promise.resolve(dispatch(createUpdateData({
    value: 2,
    timestamp: 2
  })))
}

// ====================================
// Selectors
// ====================================

export const selectData = (state: State) => state.data
export const selectThreshold = (state: State) => state.threshold

// ====================================
// Reducer
// ====================================

export const reducer: Reducer<State, Action> = (state = initState, action): State => {
  switch (action.type) {
    case UPDATE_DATA:
      return {
        ...state,
        data: [...(state.data || []), action.payload]
      }

    case UPDATE_THRESHOLD:
      return { ...state, threshold: action.payload }

    default:
      assertNever(action)
      return state
  }
}
