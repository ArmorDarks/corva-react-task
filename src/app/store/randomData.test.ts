/* eslint-env jest */
/* eslint no-unused-vars: "off" */
/* eslint @typescript-eslint/no-unused-vars: "error" */

import { reducer, initState, updateData, updateThreshold } from './randomData'

describe('randomData reducer', () => {
  describe('with no action', () => {
    it('should return same state', () => {
      const newState = reducer(undefined, {} as any)
      expect(newState).toEqual(initState)
    })
  })

  describe('with UpdateData action', () => {
    it('should change data', () => {
      const newData = { value: 5, timestamp: 232323 }
      const action = updateData(newData)

      const newState = reducer(undefined, action)

      expect(newState).toEqual({
        ...initState,
        data: [...(initState.data || []), newData]
      })
    })
  })

  describe('with UpdateThreshold action', () => {
    it('should change threshold', () => {
      const newThreshold = 5
      const action = updateThreshold(newThreshold)

      const newState = reducer(undefined, action)

      expect(newState).toEqual({
        ...initState,
        threshold: newThreshold
      })
    })
  })
})
