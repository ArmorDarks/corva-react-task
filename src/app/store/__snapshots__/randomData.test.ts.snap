// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`randomData action setConnected should create action 1`] = `
Object {
  "payload": true,
  "type": "app/randomData/SET_CONNECTED",
}
`;

exports[`randomData action updateData should create action 1`] = `
Object {
  "payload": Object {
    "timestamp": 1,
    "value": 2,
  },
  "type": "app/randomData/UPDATE_DATA",
}
`;

exports[`randomData action updateThreshold should create action 1`] = `
Object {
  "payload": 2,
  "type": "app/randomData/UPDATE_THRESHOLD",
}
`;
